src/System.IO.FileSystem.DriveInfo/tests/DriveInfo.Windows.Tests.cs

```diff
@@ -0,0 +1,194 @@
+            var validDrive = DriveInfo.GetDrives().Where(d => d.DriveType == DriveType.Fixed).First();
+            const int volNameLen = 50;
+            StringBuilder volumeName = new StringBuilder(volNameLen);
+            const int fileSystemNameLen = 50;
+            StringBuilder fileSystemName = new StringBuilder(fileSystemNameLen);
+            int serialNumber, maxFileNameLen, fileSystemFlags;
+            bool r = GetVolumeInformation(validDrive.Name, volumeName, volNameLen, out serialNumber, out maxFileNameLen, out fileSystemFlags, fileSystemName, fileSystemNameLen);
+            if (r)
+            {
+                Assert.Equal(volumeName.ToString(), validDrive.VolumeLabel);
+            }
+
+            // Set Volume Label - valid drive
+            string newLabel = "DriveInfoTests";
+            string originalLabel = validDrive.VolumeLabel;
+
+            if (validDrive.IsReady)
+            {
+                bool haveAccess = true;
+                try
+                {
+                    validDrive.VolumeLabel = newLabel;
+                    Assert.Equal(newLabel, validDrive.VolumeLabel);
+                }
+                catch (UnauthorizedAccessException) { haveAccess = false; }
+                finally
+                {
+                    if (haveAccess)
+                        validDrive.VolumeLabel = originalLabel;
+                }
+            }
+
+            // Test UnauthorizedAccess on Network or CD-ROM
+            var noAccessDrive = DriveInfo.GetDrives().Where(d => d.DriveType == DriveType.Network || d.DriveType == DriveType.CDRom);
+            foreach (var adrive in noAccessDrive)
+            {
+                if (adrive.IsReady)
+                    Assert.Throws<UnauthorizedAccessException>(() => { adrive.VolumeLabel = null; });
+            }
+
+            // Test Invalid drive
+            var invalidDrive = new DriveInfo(GetInvalidDriveLettersOnMachine().First().ToString());
+            Assert.Throws<DriveNotFoundException>(() => { var vl = invalidDrive.VolumeLabel; });
+            Assert.Throws<DriveNotFoundException>(() => { invalidDrive.VolumeLabel = null; });
+        }
+
+        [DllImport("kernel32.dll", SetLastError = true)]
+        internal static extern int GetLogicalDrives();
+
+        [DllImport("kernel32.dll", SetLastError = true)]
+        internal static extern bool GetVolumeInformation(string drive, StringBuilder volumeName, int volumeNameBufLen, out int volSerialNumber, out int maxFileNameLen, out int fileSystemFlags, StringBuilder fileSystemName, int fileSystemNameBufLen);
+
+        [DllImport("kernel32.dll", SetLastError = true)]
+        internal static extern int GetDriveType(string drive);
+
+        private IEnumerable<char> GetValidDriveLettersOnMachine()
+        {
+            uint mask = (uint)GetLogicalDrives();
+            if (mask == 0)
+            {
+                Assert.False(true, "Test Failed, Error calling Win32 GetLogicalDrive.");
```

Comment:
I know there was discussion in the past about how to best log messages from asserts and this is only possible with Assert.False/True, but personally I don't like this way. From reading various places it seems the suggestion is to just throw an exception instead. Was this discussed as an alternative?